### ISSUE TITLE
C++ overloaded operators break if no nullary constructor available

### ISSUE THREAD

**Issue by dgrell** [cython/cython#1465]
Since version 0.24, the code generated by cython for overloaded `operator+` and `operator-` seems to have changed. The result of the operation used to be held in a temporary pointer before. Now, the temporary object is constructed directly on the stack, requiring a nullary constructor to exist. Compilation fails if no such constructor is available:

Source:

```
cdef c.Bin1D_Dbn1D* c_ptr = new c.Bin1D_Dbn1D(  deref(self.ptr()) + deref(other.ptr())  )
```

Cython 0.23.4:

```
  __pyx_t_4yoda_12declarations_Bin1D_Dbn1D *__pyx_t_3;
// ...
try {
    __pyx_t_3 = new __pyx_t_4yoda_12declarations_Bin1D_Dbn1D(((*__pyx_t_1) + (*__pyx_t_2)));
  } catch(...) {
```

Cython 0.24.1:

```
YODA::Bin1D<YODA::Dbn1D>  __pyx_t_3;
// ...
try {
    __pyx_t_3 = ((*__pyx_t_1) + (*__pyx_t_2));
  } catch(...) {
```

The Bin1D<> class does not have a nullary constructor available, and thus compilation fails.

Looking at the changelog, it may be related to this item in the bug fixes for 0.24:

> C++ exceptions raised by overloaded C++ operators were not always handled. Patch by Ian Henriksen.

**Comment by insertinterestingnamehere** [comment]
This is a limitation that's pretty inherent to the way Cython does a lot of its code generation. The fact that it worked before was primarily due to buggy forwarding of exceptions from C++ to Python. If you want the old behavior (where exceptions thrown in overloaded operators aren't guaranteed to be converted to Python exceptions) it should probably be enough to remove `except +` from your `extern` declarations for the overloaded operators.
This isn't really possible with the current exception forwarding model though. In order to avoid calling the nullary constructor, Cython would have to create a variable inside a try block (try-catch idioms are used to forward C++ exceptions) that would exist once the try-catch block had finished. C++ doesn't allow that.
In theory, one way to work around stuff like this would be to introduce a feature that allows multiple function calls, etc. to be handled within the same code try-catch statement. That way you could create and use the temporary variable that you need all within the same scope at the C++ level. Syntactically something like that would make sense as some sort of `with` statement. That's a feature I'd love to see at some point, but I haven't had the time to dig into it much.

**Comment by dgrell** [comment]
Thanks for your comments! I wasn't concerned with the exception behaviour originally, and still don't fully see how the new better functionality requires the syntactical change since 0.24, where Cython issues code like

```
Foo * a = ...;
Foo * b = ...;
Foo tmp;
// ...
try {
  tmp = *a + *b;
} 
```

instead of what used to be effectively

```
Foo * a = ...;
Foo * b = ...;
Foo * tmp;
// ...
try {
  tmp = new Foo(*a + *b);
}
```
which as far as I can see won't make a difference to exceptions thrown by operator+()

[The docs](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html) also suggest using pointers for cases where there is no nullary constructor:

> If the class you’re wrapping does not have a nullary constructor, you must store a pointer to the wrapped class [...]

Right now, Cython 0.24 makes it impossible to use such classes at all, and most of the interface packages I maintain are broken. I'd like to think that that was accidental, and not a conscious decision to disallow classes that don't have nullary constructors.

If I'm missing the point here, can you please explain again why the new exception behaviour is not possible with the old pointer pattern? Thanks!

**Comment by insertinterestingnamehere** [comment]
Right, the current limitation doesn't have much to do with arithmetic operators in particular. Really any function or arithmetic operation that has an exception handler can't return something that doesn't have a default constructor because it has to allocate a temporary to assign to before running the try-catch block. Prior to the change, code like `c = a + b` wouldn't call the exception handler at all, and instead the interpreter would crash when an uncaught C++ exception was thrown. Fixing that meant requiring slightly more of classes that use exception handling with arithmetic operators, but the behavior was broken to begin with.

The reason the separate try-catch block for the arithmetic operation is necessary at all is that the function call and the overloaded operator can use completely different custom exception handlers. The same would be true if the argument passed to the function with the exception handler were just the result of another function with its own exception handler. Either way, for each subexpression, Cython has to create a new temporary in the surrounding scope and then assign the result of the function call to it in a separate try-catch block. The assignment has to be to a variable outside the try block because it needs to persist after the try-catch block finishes. The assignment has to be by value as well, otherwise all that would be left is an invalid address pointing to an already deallocated temporary value.

In theory it's possible roll together multiple function calls and arithmetic subexpressions into a single try-catch statement so that classes without default constructors can be used as temporaries, but fusing the exception handling like that is a not a trivial feature. It just happens to have been the net result of the buggy behavior in the past. A change like that would also involve some changes in semantics that would make forwarded exceptions come from less specific areas of code. Right now, exceptions should be able to point precisely to the function call or arithmetic operation in Cython that caused them. It's not really clear how to choose which subexpressions (or which lines) should have their try-catch blocks combined either. That's why I mentioned that, if anyone ever implements fusing try-catch statements like this, it should probably be enabled using some sort of `with` block.

**Comment by scopatz** [comment]
For what it is worth, I hunted down this issue for pyjsoncpp and the removal of `except +` did seem to fix it.  However, this could really use some documentation somewhere.  I may be misreading something, but it didn't look like the changelog mentioned this at all.  Thanks for your efforts on Cython @insertinterestingnamehere, but I did spend the better part of a day tracking this down.

**Comment by dgrell** [comment]
I can confirm that removing `except +` works around my problem, too. I agree with @scopatz that this needs to be documented much more prominently, since the compile error gives no hints that the issue is related to exception passing in operators. To summarize in one line for future readers here, 

_In Cython 0.24, using `Foo::operatorX()` with `except +` requires the existence of a useable nullary constructor on `Foo`._
